File: android/app/src/main/java/com/kryptos/android/core/TextStego.kt
package com.kryptos.android.core

import java.security.SecureRandom
import java.text.Normalizer
import java.util.Locale

enum class StegoLanguage {
    ENGLISH, RUSSIAN;

    val words: List<String>
        get() = when (this) {
            ENGLISH -> Wordlists.english
            RUSSIAN -> Wordlists.russian
        }

    internal val indexMap: Map<String, Int>
        get() = when (this) {
            ENGLISH -> Wordlists.englishIndex
            RUSSIAN -> Wordlists.russianIndex
        }

    companion object {
        fun forSystem(): StegoLanguage =
            if (Locale.getDefault().language.lowercase().startsWith("ru")) RUSSIAN else ENGLISH
    }
}

internal object Wordlists {
    val english: List<String> by lazy { load("english") }
    val russian: List<String> by lazy { load("russian") }
    val englishIndex: Map<String, Int> by lazy { index(english) }
    val russianIndex: Map<String, Int> by lazy { index(russian) }

    private fun load(name: String): List<String> {
        val stream = Wordlists::class.java.classLoader!!.getResourceAsStream("wordlists/$name.txt")
            ?: error("wordlist $name missing from resources")
        val words = stream.bufferedReader(Charsets.UTF_8).readLines().filter { it.isNotBlank() }
        check(words.size == 4096) { "word list must hold exactly 4096 words, got ${words.size}" }
        return words
    }

    private fun index(words: List<String>): Map<String, Int> =
        words.withIndex().associate { (i, w) -> Normalizer.normalize(w, Normalizer.Form.NFC) to i }
}

object TextStego {
    private const val BITS_PER_WORD = 12
    private const val WORD_MASK = 0xFFF
    private const val MAGIC = 0xC7

    const val MAX_PAYLOAD_BYTES = 0x7FFF

    private val random = SecureRandom()

    fun encode(data: ByteArray, language: StegoLanguage = StegoLanguage.forSystem()): String =
        encode(data, language, random.nextInt(256))

    internal fun encode(data: ByteArray, language: StegoLanguage, seed: Int): String {
        val words = language.words
        require(data.size <= MAX_PAYLOAD_BYTES) { "payload too large for the stego frame" }

        val shortLen = data.size <= 0x7F
        val headerSize = if (shortLen) 2 else 3
        val inner = ByteArray(headerSize + data.size + 1)
        inner[0] = MAGIC.toByte()
        if (shortLen) {
            inner[1] = data.size.toByte()
        } else {
            inner[1] = (0x80 or (data.size shr 8)).toByte()
            inner[2] = (data.size and 0xFF).toByte()
        }
        data.copyInto(inner, headerSize)
        inner[inner.size - 1] = crc8(inner, 0, inner.size - 1).toByte()

        val framed = ByteArray(1 + inner.size)
        framed[0] = seed.toByte()
        var x = seed and 0xFF
        for (i in inner.indices) {
            x = (x * 197 + 91) and 0xFF
            framed[1 + i] = (inner[i].toInt() xor x).toByte()
        }

        val indices = ArrayList<Int>((framed.size * 8 + BITS_PER_WORD - 1) / BITS_PER_WORD)
        var acc = 0
        var bits = 0
        for (byte in framed) {
            acc = (acc shl 8) or (byte.toInt() and 0xFF)
            bits += 8
            while (bits >= BITS_PER_WORD) {
                bits -= BITS_PER_WORD
                indices.add((acc shr bits) and WORD_MASK)
            }
        }
        if (bits > 0) indices.add((acc shl (BITS_PER_WORD - bits)) and WORD_MASK)

        return prettify(indices.map { words[it] }, seed)
    }

    fun decode(text: String): ByteArray? {
        val tokens = tokenize(text)
        if (tokens.isEmpty()) return null
        for (language in StegoLanguage.entries) {
            decode(tokens, language)?.let { return it }
        }
        return null
    }

    fun looksLikeStego(text: String): Boolean = decode(text) != null

    fun mightBeStego(text: String, sample: Int = 6): Boolean {
        val tokens = tokenize(text.take(200))
        if (tokens.size < sample) return false
        val head = tokens.take(sample * 2)
        return StegoLanguage.entries.any { lang -> head.count { lang.indexMap.containsKey(it) } >= sample }
    }

    private fun decode(tokens: List<String>, language: StegoLanguage): ByteArray? {
        val index = language.indexMap
        var acc = 0L
        var bits = 0
        val bytes = ArrayList<Byte>(tokens.size * BITS_PER_WORD / 8 + 1)
        for (token in tokens) {
            val value = index[token] ?: continue
            acc = (acc shl BITS_PER_WORD) or value.toLong()
            bits += BITS_PER_WORD
            while (bits >= 8) {
                bits -= 8
                bytes.add(((acc shr bits) and 0xFF).toByte())
            }
        }
        if (bytes.size < 4) return null
        var x = bytes[0].toInt() and 0xFF
        val inner = ByteArray(bytes.size - 1)
        for (i in inner.indices) {
            x = (x * 197 + 91) and 0xFF
            inner[i] = (bytes[1 + i].toInt() xor x).toByte()
        }
        if ((inner[0].toInt() and 0xFF) != MAGIC) return null
        val b1 = inner[1].toInt() and 0xFF
        val n: Int
        val offset: Int
        if (b1 < 0x80) {
            n = b1
            offset = 2
        } else {
            if (inner.size < 3) return null
            n = ((b1 and 0x7F) shl 8) or (inner[2].toInt() and 0xFF)
            offset = 3
        }
        if (inner.size < offset + n + 1) return null
        if ((inner[offset + n].toInt() and 0xFF) != crc8(inner, 0, offset + n)) return null
        return inner.copyOfRange(offset, offset + n)
    }

    private fun crc8(data: ByteArray, from: Int, to: Int): Int {
        var crc = 0
        for (i in from until to) {
            crc = crc xor (data[i].toInt() and 0xFF)
            repeat(8) {
                crc = if (crc and 0x80 != 0) ((crc shl 1) xor 0x07) and 0xFF else (crc shl 1) and 0xFF
            }
        }
        return crc
    }

    private fun prettify(words: List<String>, seed: Int): String {
        if (words.isEmpty()) return ""
        var x = (seed xor 0xA5) and 0xFF
        fun next(): Int {
            x = (x * 197 + 91) and 0xFF
            return x
        }
        val sentences = ArrayList<String>()
        var i = 0
        while (i < words.size) {
            val len = 4 + next() % 6
            val chunk = words.subList(i, minOf(i + len, words.size))
            val sb = StringBuilder()
            for ((k, word) in chunk.withIndex()) {
                if (k > 0) sb.append(' ')
                sb.append(word)
                if (k < chunk.size - 1 && next() % 6 == 0) sb.append(',')
            }
            val mark = when (next() % 10) {
                8 -> "?"
                9 -> "!"
                else -> "."
            }
            sentences.add(sb.toString().replaceFirstChar { it.uppercase() } + mark)
            i += len
        }
        return sentences.joinToString(" ")
    }

    private fun tokenize(text: String): List<String> = StegoTokenizer.split(text)
}

File: CipherCore/Sources/CipherCore/Steganography/TextStego.swift
import Foundation

public enum StegoLanguage: String, Sendable, CaseIterable {
    case english
    case russian

    var words: [String] {
        switch self {
        case .english: return StegoWordlists.english
        case .russian: return StegoWordlists.russian
        }
    }

    public static func forSystem() -> StegoLanguage {
        let code = (Locale.preferredLanguages.first ?? "en").lowercased()
        return code.hasPrefix("ru") ? .russian : .english
    }
}

public enum TextStego {
    private static let bitsPerWord = 12
    private static let wordMask = 0xFFF
    private static let magic: UInt8 = 0xC7

    public static let maxPayloadBytes = 0x7FFF

    public static func encode(_ data: Data, language: StegoLanguage = .forSystem()) -> String {
        encode(data, language: language, seed: Int.random(in: 0 ... 255))
    }

    static func encode(_ data: Data, language: StegoLanguage, seed: Int) -> String {
        let words = language.words
        precondition(words.count == 1 << bitsPerWord, "word list must hold exactly 4096 words")
        precondition(data.count <= maxPayloadBytes, "payload too large for the stego frame")

        var inner = [UInt8]()
        inner.reserveCapacity(4 + data.count)
        inner.append(magic)
        if data.count <= 0x7F {
            inner.append(UInt8(data.count))
        } else {
            inner.append(UInt8(0x80 | (data.count >> 8)))
            inner.append(UInt8(data.count & 0xFF))
        }
        inner.append(contentsOf: data)
        inner.append(crc8(inner, 0, inner.count))

        var framed = [UInt8]()
        framed.reserveCapacity(1 + inner.count)
        framed.append(UInt8(seed & 0xFF))
        var x = seed & 0xFF
        for byte in inner {
            x = (x * 197 + 91) & 0xFF
            framed.append(byte ^ UInt8(x))
        }

        var indices: [Int] = []
        indices.reserveCapacity((framed.count * 8 + bitsPerWord - 1) / bitsPerWord)
        var acc = 0, bits = 0
        for byte in framed {
            acc = (acc << 8) | Int(byte)
            bits += 8
            while bits >= bitsPerWord {
                bits -= bitsPerWord
                indices.append((acc >> bits) & wordMask)
            }
        }
        if bits > 0 { indices.append((acc << (bitsPerWord - bits)) & wordMask) }

        return prettify(indices.map { words[$0] }, seed: seed)
    }

    public static func decode(_ text: String) -> Data? {
        let tokens = tokenize(text)
        guard !tokens.isEmpty else { return nil }
        for language in StegoLanguage.allCases {
            if let data = decode(tokens: tokens, language: language) { return data }
        }
        return nil
    }

    public static func looksLikeStego(_ text: String) -> Bool { decode(text) != nil }

    private static func decode(tokens: [String], language: StegoLanguage) -> Data? {
        let index = language.indexMap
        var acc = 0, bits = 0
        var bytes: [UInt8] = []
        bytes.reserveCapacity(tokens.count * bitsPerWord / 8 + 1)
        for token in tokens {
            guard let value = index[token] else { continue }
            acc = (acc << bitsPerWord) | value
            bits += bitsPerWord
            while bits >= 8 {
                bits -= 8
                bytes.append(UInt8((acc >> bits) & 0xFF))
            }
        }
        guard bytes.count >= 4 else { return nil }
        var x = Int(bytes[0])
        var inner = [UInt8]()
        inner.reserveCapacity(bytes.count - 1)
        for i in 1 ..< bytes.count {
            x = (x * 197 + 91) & 0xFF
            inner.append(bytes[i] ^ UInt8(x))
        }
        guard inner[0] == magic else { return nil }
        let b1 = Int(inner[1])
        let n: Int
        let offset: Int
        if b1 < 0x80 {
            n = b1
            offset = 2
        } else {
            guard inner.count >= 3 else { return nil }
            n = ((b1 & 0x7F) << 8) | Int(inner[2])
            offset = 3
        }
        guard inner.count >= offset + n + 1 else { return nil }
        guard inner[offset + n] == crc8(inner, 0, offset + n) else { return nil }
        return Data(inner[offset ..< offset + n])
    }

    private static func crc8(_ data: [UInt8], _ from: Int, _ to: Int) -> UInt8 {
        var crc = 0
        for i in from ..< to {
            crc ^= Int(data[i])
            for _ in 0 ..< 8 {
                crc = (crc & 0x80) != 0 ? ((crc << 1) ^ 0x07) & 0xFF : (crc << 1) & 0xFF
            }
        }
        return UInt8(crc)
    }

    private static func prettify(_ words: [String], seed: Int) -> String {
        guard !words.isEmpty else { return "" }
        var x = (seed ^ 0xA5) & 0xFF
        func next() -> Int {
            x = (x * 197 + 91) & 0xFF
            return x
        }
        var sentences: [String] = []
        var i = 0
        while i < words.count {
            let len = 4 + next() % 6
            let chunk = Array(words[i ..< min(i + len, words.count)])
            var sentence = ""
            for (k, word) in chunk.enumerated() {
                if k > 0 { sentence += " " }
                sentence += word
                if k < chunk.count - 1, next() % 6 == 0 { sentence += "," }
            }
            let mark: String
            switch next() % 10 {
            case 8: mark = "?"
            case 9: mark = "!"
            default: mark = "."
            }
            sentence = sentence.prefix(1).uppercased() + sentence.dropFirst()
            sentences.append(sentence + mark)
            i += len
        }
        return sentences.joined(separator: " ")
    }

    private static func tokenize(_ text: String) -> [String] {
        let normalized = text.precomposedStringWithCanonicalMapping.lowercased()
        return normalized
            .split(whereSeparator: { !$0.isLetter })
            .map(String.init)
    }
}

private extension StegoLanguage {
    static let indexMaps: [StegoLanguage: [String: Int]] = {
        var maps = [StegoLanguage: [String: Int]](minimumCapacity: allCases.count)
        for language in allCases {
            var map = [String: Int](minimumCapacity: 4096)
            for (i, word) in language.words.enumerated() {
                map[word.precomposedStringWithCanonicalMapping] = i
            }
            maps[language] = map
        }
        return maps
    }()

    var indexMap: [String: Int] { Self.indexMaps[self]! }
}
